### ISSUE TITLE
JdbcSelect hangs pipeline when doing more than 8 concurrent selects

### ISSUE THREAD

**Issue by muscovitebob** [spotify/scio#2774]
I have a project that closely follows the example [here](https://github.com/spotify/scio/blob/master/scio-examples/src/main/scala/com/spotify/scio/examples/extra/CloudSqlExample.scala), but I query multiple tables with same `select *`. I end up having a list of SCollections that I then do unionAll on to get all of these results into a single thing I can work with neatly from there on.

However, if I try to query more than 8 tables (hence I will have more than 8 separate Dataflow steps that all get combined in the unionAll), the pipeline simply freezes until it times out. Like the example assumes, I am querying a MySQL database via the proxy, hosted in CloudSQL.
 I observe the same behaviour on my direct runner. At or below 8 selects, everything works.

I have made sure it is not some setting in the MySQL DB related to concurrency or permissions, but this should be relatively easy to reproduce from my description.

**Comment by jto** [comment]
Hi @muscovitebob, 

Thanks for reporting this :)

Scio is merely wrapping beam's `JdbcIO` so it's likely a bug in Beam itself. 
The symptom sound like thread starvation. 

Could you get a thread dump while this it is happening ?

**Comment by muscovitebob** [comment]
Hi @jto, indeed I have a thread dump from such a frozen Beam job. I wasn't able to glean much actionable insight from it but if you can I would really appreciate it!

[thread_dump_50967.log](https://github.com/spotify/scio/files/4349417/thread_dump_50967.log)

**Comment by muscovitebob** [comment]
To give some further insight on this: whatever freezes the job must happen even before `JdbcIO` does any calls to the database, as I had a job with invalid SQL syntax idle until its timeout value, and only detect the syntax was invalid once dataflow started shutting down the workers.

**Comment by jto** [comment]
Hi @muscovitebob,

I think what may be happening is that all the connections on the connection pool get used, which lead to `BasicDataSource.getConnection` to just block.

I suspect that Dataflow waits for the `setup` of each `DoFn` instance to be completed before it proceeds to any actual execution. Since some `getConnection` are blocked, everything get stuck. Here's the relevant bit in your dump:

```java
at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at org.apache.commons.pool2.impl.LinkedBlockingDeque.takeFirst(LinkedBlockingDeque.java:590)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:425)
	at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:346)
	at org.apache.commons.dbcp2.PoolingDataSource.getConnection(PoolingDataSource.java:134)
	at org.apache.commons.dbcp2.BasicDataSource$PaGetConnection.run(BasicDataSource.java:73)
	at org.apache.commons.dbcp2.BasicDataSource$PaGetConnection.run(BasicDataSource.java:69)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:800)
	at org.apache.beam.sdk.io.jdbc.JdbcIO$ReadFn.setup(JdbcIO.java:881)
```

Could you open an issue with Dataflow and link it here ?

**Comment by muscovitebob** [comment]
Thanks very much for looking into this @jto!

Since I also see this on Direct Runner and from the dump fragment the blocking is in a beam library I am assuming you meant filing an issue on the Apache Beam issue tracker, which I will do.

**Comment by jto** [comment]
Yep indeed :) Thanks.

**Comment by jto** [comment]
Hi @muscovitebob,

did you open an issue with Beam ? 
Could you link it here ?

Thanks :)

**Comment by muscovitebob** [comment]
Hi @jto,

My apologies, we've switched to working on another project and I've only just got around to filing a bug at Beam. You can find it [here](https://issues.apache.org/jira/browse/BEAM-9629).

Thanks for your help again,

Boris

**Comment by muscovitebob** [comment]
Just a heads up, looks like this has been fixed in Beam now.

**Comment by nevillelyh** [comment]
Thanks. Ticket says fixed in 2.24.0, which we'll upgrade to once it's released.

**Comment by regadas** [comment]
Closed via #3325
